Modification to GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc

rsadikin opened the pull request:
Unit test updated due to changes in local distortions

    Add correction with side information

    ATO 112 - fixed for irregular interpolator desctruction, correction

    ATO 112: fix drawing bugs for DistDR

    Spacecharge Test accomodate TF2

shahor02 commented:
@rsadikin Could you please check the (red) results of the CI test (the o2-dev-fairroot can be ignored). Eventually the code should be passed via clang-format (see https://github.com/AliceO2Group/CodingGuidelines). I see that in some cases your modifications are regressive, e.g. you substitute the new style headers like ``cstdlib`` by old ``stdlib.h``. This is interpreted as an error by CI.

rsadikin commented:
> @rsadikin Could you please check the (red) results of the CI test (the o2-dev-fairroot can be ignored). Eventually the code should be passed via clang-format (see https://github.com/AliceO2Group/CodingGuidelines). I see that in some cases your modifications are regressive, e.g. you substitute the new style headers like `cstdlib` by old `stdlib.h`. This is interpreted as an error by CI.

Thanks for your comments. Sorry for errors, I need to check it locally first before make a commit. Should I rebase to dev branch before make a new commit?

davidrohr commented:
Yes, please rebase to dev branch first (as there were many changes), and the address the comments.

ehellbar commented:
Dear Rifki,

can you please keep the curly brackets in for / if statements?
It should be like
`for/if (...){
...
}`

rsadikin commented:
> Dear Rifki,
> 
> can you please keep the curly brackets in for / if statements?
> It should be like
> for/if (...){
> ...
> }

Ok, I will do it, I think it is because of clang-format execution https://github.com/AliceO2Group/CodingGuidelines
I will reset and git push again.

rsadikin commented:
Dear @shahor02,  @davidrohr and @ehellbar 

I think I already fixed some issues regarding clang-format, and code style format. I also squashed all commit to one.

Please your advice.

Best Regards
Rifki Sadikin

davidrohr commented on GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.h:
```diff
@@ -440,8 +528,7 @@ class AliTPCSpaceCharge3DCalc
   /// \cond CLASSIMP
-  ClassDef(AliTPCSpaceCharge3DCalc,
-           1);
+  ClassDef(AliTPCSpaceCharge3DCalc, 1);
```
for non-virtual classes you can use ClassDefNV

davidrohr commented on GPU/TPCSpaceChargeBase/AliTPC3DCylindricalInterpolatorIrregular.cxx:
```diff
@@ -622,6 +623,10 @@ void AliTPC3DCylindricalInterpolatorIrregular::SetValue(
 {
   Int_t indexInner;
   Int_t index;
+  TMatrixD* mat;
```
I don't understand this change? The scope of the variables can remain in the for loop, as suggested by Codacy. Or do I miss something?

ehellbar commented on GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.cxx:
```diff
@@ -31,6 +31,18 @@ ClassImp(AliTPCSpaceCharge3DCalc)
+    fNZColumns(129),
+    fNPhiSlices(180),
+    fCorrectionType(1),
+    fIntegrationStrategy(0)
```
Dear Rifki,

this information is redundant as the default values for these variables are already defined in the header file.

This is an artefact of the old AliTPCCommon code and this happens at several points in your pull request. Before the merge of AliTPCCommon into O2/GPU, David and I went trough the code to make it compatible with modern coding guidelines and the clang-format syntax. In your pull request, you reverse some of these changes.

I would kindly ask you to make a clean rebase to O2/dev, only import the actual changes that you did since this version and make a new pull request with only these changes. There should be no fundamental format or syntax / coding structure changes of the current code version in the O2/dev branch because it already was brought up to date before the merge.

I suggest to not merge and close this pull request.

Cheers,
Ernst

rsadikin replied:
Hi Ernst

Thanks for your comments.
I will do as you requested, made changes on @dev, and make a new pull request.
Sorry I didn't realize that you have made a lot changes.

Regards,
Rifki Sadikin

rsadikin commented:
Hi @ehellbar and all,

Apologise for delay on update. Here I made several changes related to CPU-version of SpaceCharge at GPU/TPCSpaceCharge directory. I am based the code from dev branch, and only add changes I made:


1. On pointer deletion in AliTPC3DCylindricalInterpolatorIrregular, destructor made an error. 
2. On Correction look up table for irregular grid, the value of r-correction should be negative of r- distortion.
3. TF1 accepting TF2 (since TF2 is the child of TF1) type castings are possible for boundary function.
4. InitSpaceCharge3DDriftLine for analytic functions. Given all analytic functions, the implementation could calculate distortions.
5. I provide new test function and AliTPCSpaceCharge3DDriftline interface on AliRoot (will be pulled request) after this AliceO2 pull request.
6. This pull request, supposed to not giving an error when merge to AliRoot. This should be run well under the current AliRoot/TPC/TPCbase/test

I attached 2 files of ctest output, the first one is ctest on the current AliRoot master see [output-master-ctest.txt](https://github.com/AliceO2Group/AliceO2/files/3491410/output-master-ctest.txt),  the second one is the ctest output on the update AliRoot (point 5.) 
[output-after-o2-ctest.txt](https://github.com/AliceO2Group/AliceO2/files/3491417/output-after-o2-ctest.txt)


Best Regards
Rifki Sadikin

ehellbar commented on GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.cxx:
```diff
@@ -1401,29 +1440,21 @@ void AliTPCSpaceCharge3DCalc::InitSpaceCharge3DPoissonIntegralDz(
-        f1BoundaryOFC = fFormulaBoundaryOFCC;
-        f1BoundaryROC = fFormulaBoundaryROCC;
+        f1BoundaryOFC = fFormulaBoundaryOFCA;
+        f1BoundaryROC = fFormulaBoundaryROCA;
```
Shouldn't these be the boundary formulas for C side since it is in the C side part of the condition?

ehellbar commented on GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.cxx:
```diff
@@ -2242,8 +2356,9 @@ void AliTPCSpaceCharge3DCalc::IntegrateDistCorrDriftLineDz(
+          // values should be negative of distortions
           (*mCorrIrregularDrDz)(i, j) = -drDist;
-          (*mCorrIrregularDPhiRDz)(i, j) = -1 * dPhi * (radius0 + drDist);
+          (*mCorrIrregularDPhiRDz)(i, j) = -1 * dPhi * radius0;
```
The reference points for the irregular correction lookup tables are the distorted points, right? So for the drphi correction, it should be dPhi * the distorted radius, which is radius0 + drDist.
(*mCorrIrregularDPhiRDz)(i, j) = -1 * dPhi * (radius0 + drDist);

ehellbar commented on GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.cxx:
```diff
@@ -2453,32 +2562,33 @@ void AliTPCSpaceCharge3DCalc::IntegrateDistCorrDriftLineDz(
+        // use oppsite directions of distortion
         (*mCorrIrregularDrDz)(i, j) = -drDist;
-        (*mCorrIrregularDPhiRDz)(i, j) = -dRPhi;
+        (*mCorrIrregularDPhiRDz)(i, j) = -dPhi * radius0;
```
Same thing. As you use the distorted points as the reference points for the irregular correction lookup tables, the correction in rphi is -1 times distortion in phi (-dPhi) times the distorted radius (radius0 + drDist)

ehellbar commented on GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.cxx:
```diff
@@ -2698,9 +2809,9 @@ void AliTPCSpaceCharge3DCalc::IntegrateDistCorrDriftLineDz(
+        // use oppsite directions of distortion
         (*mCorrIrregularDrDz)(i, j) = -drDist;
-        (*mCorrIrregularDPhiRDz)(i, j) = -dPhi * (radius0 + drDist);
+        (*mCorrIrregularDPhiRDz)(i, j) = -dPhi * radius0;
```
Same thing using the distorted points as reference points for the irregular correction lookup tables. Should be 
(*mCorrIrregularDPhiRDz)(i, j) = -1 * dPhi * (radius0 + drDist);

ehellbar commented:
Other than the few comments above, the PR looks fine to me.

davidrohr commented:
I merged this locally and tried also with AliRoot. Looks good.
Can be merged once the points @ehellbar stated are fixed.

rsadikin commented:
Hello @davidrohr 
I saw a new code at the last, which you modify the codes  to conform codacy standards such as introducting kMaxPhiSlices and so on.
Would you please put it back to dev-branch and I will make modification accordingly.

Hi @ehellbar, I made changes as you suggest on Phi Correction, however it made the testing at AliRoot little bit wrong since at the procedure GenerateDistortionTree at AliTPCSpaceCharge3DDriftline at AliRoot, assumption Phi Correction using original R that's why it did't meet. 
After this pull request merged, I will also need to update AliTPCSpaceCharge3DDriftline at AliRoot

Best Regards
Rifki Sadikin

davidrohr commented:
Hello @rsadikin : I do not fully understand your question. Could you please clarify what you mean?
In general:
- The GPU and the TPC SCD should be developed inside the O2 repository (dev branch). My latest changes are all in this dev branch already.
- The GPU folder should be synchronized over to AliRoot regularly. I am a bit lacking behind here, sorry for that, I'll sync the current state over as soon as possible.
- For testing the Software with AliRoot, you can also just copy over the full GPU folder to AliRoot locally.

sawenzel commented:
This is getting out of date. What is the status here?

ehellbar commented:
Hello @sawenzel , @rsadikin ,
I will review the changes with respect to the latest questions this week and let you know if we can proceed here.

davidrohr commented:
@rsadikin : Could you adress @ehellbar 's changes, such that we can get this merged eventually?

ktf commented:
Is this still worked on?

ehellbar commented:
Hello @ktf , 

can we keep the PR as work in progress for the time being? I hope @rsadikin will take care of the suggested changes soon. 

In any case, I will open another PR with more modifications of the same code soon. If the status of this one will not have changed in the meantime, I will incorporate the changes of this PR into mine.

Is this ok for you?

Cheers,
Ernst

ehellbar commented:
Hello @ktf , @davidrohr , @rsadikin ,

I adopted the modifications from this PR to my recent one (https://github.com/AliceO2Group/AliceO2/pull/3048). Please close this PR without merging.

Thanks and cheers,
Ernst
